c/tskit/trees.h

```diff
@@ -229,49 +229,50 @@ int tsk_treeseq_loadf(tsk_treeseq_t *self, FILE *file, tsk_flags_t options);
 int tsk_treeseq_dump(tsk_treeseq_t *self, const char *filename, tsk_flags_t options);
```

Comment:
Both dump methods can have ``const self`` I think. In general, the tsk_treeseq_t class is immutable, so nearly every method should have have a ``const`` self. Other than, I think, ``init``, ``load``, and ``free``.

Reply:
I was worried about this TSK_NO_BUILD_INDEXES option and rebuilding of indexes within the implementation of dump. I think maybe when data inside is not sorted. It's hazy to me, I'm so new to it.

I've come across the terms 'physical state' and 'logical state': https://isocpp.org/wiki/faq/const-correctness#logical-vs-physical-const
So a question is whether this potential rebuilding of indexes is totally "invisible" and thus just change in "physical" state and not "logical" state. My inclination would be, in C, to avoid this kind of advanced subtly that probably makes more sense in C++, where data members can be private. But in C is harder to be sure somebody might not be "seeing" changes in "physical" state even if it's not change in "logical" state.

Is this rebuilding of indexes not possible in this TS dump?

Reply:
In my mind the indexes are part of the logical state of a `TableCollection` as they are exposed, and not an internal implementation detail.  In the current state then `dump` cannot be `const`.

Do we break much if we remove the option to build indexes on dump? It seems a bit un-intuative that dumping creates the index.

Reply:
Great points, thanks both. I think dump can probably have the option to build the indexes removed, as this is left over from an earlier version of the code that didn't have such good support for indexes being present/not present. However, this is too much to do in this PR - let's open an issue to track tagging ``tsk_treeseq_dump`` as const and removing this option.

Reply:
OK, resolved for this PR. I'll create an issue on this unless I hear otherwise.